website/docs/tutorials/react-rsc.md

```diff
@@ -0,0 +1,198 @@
+:::
+
+Lastly, there's the `appRouterI18n.ts` file, which is only executed on server and holds one instance of I18n object for each locale of our application. See [here](#) how it's implemented in the example.
+
+### Rendering translations in server and client components
+
+Below you can see an example of a React component. This component can be rendered **both with RSC and on client**. This is great if you're migrating a Lingui-based project from pages router to app router because you can keep the same components working in both worlds.
+
+In fact, if you swapped the html tags for their more universal alternatives, this component could also be used in React Native.
+
+```tsx title="app/[lang]/components/SomeComponent.tsx"
+import { Trans, t } from "@lingui/macro";
+
+export function SomeComponent() {
+  const { i18n } = useLingui();
+  return (
+    <div>
+      <p>
+        <Trans>Some Item</Trans>
+      </p>
+      <p>{t(i18n)`Other Item`}</p>
+    </div>
+  );
+}
+```
+
+As you might recall, hooks are not supported in RSC, so you might be surprised that this works. Under RSC, `useLingui` is actually not a hook but a simple function call which reads from the React `cache` mentioned above.
+
+Here's how you could implement server-side `useLingui` yourself using `getI18n`, which is another way to obtain the I18n instance on the server:
+
+```tsx title="useLinguiForRSC.ts"
+import { getI18n } from "@lingui/react/server";
+
+export function useLingui() {
+  const ctx = getI18n();
+  if (!ctx) {
+    // throw an informative error
+  }
+
+  return ctx;
+}
+```
+
+### Pages, Layouts and Lingui
+
+There's one last caveat: in a real-world app, you will need to localize many pages, and layouts. Because of the way the app router is designed, the `setI18n` call needs to happen not only in layouts, but also in pages. Read more in:
+
+- [Why do nested layouts/pages render before their parent layouts?](https://github.com/vercel/next.js/discussions/53026)
+- [On navigation, layouts preserve state and do not re-render](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#layouts)
+
+This means you need to repeat the `setI18n` in every page and layout. Luckily, you can easily factor it out into a simple function call, or create a HOC with which you'll wrap pages and layouts [as seen here](#). Please let us know if there's a known better way.
+
+### Changing the active language
+
+Most likely, your users will not need to change the language of the application because it will render in their preferred language (obtained from the `accept-language` header in the [middleware](#)), or with a fallback.
+
+To change language, you can either redirect users to a page with the new locale in the url, or you can load another locale [dynamically](/guides/dynamic-loading-catalogs.md). However, you need to keep in mind that if you have server-rendered locale-dependent content, it'd become stale.
```

Comment:
no sure we need to advertise dynamic catalogs loading here. For the nextjs the only and single correct way would be to simply navigate to page with correct language, due to nextjs hybrid router + RSC, the only needed parts whould be updated, so pretty much the same result as if we load it dynamically on the FE

Reply:
yeah good point. I rewrote it slightly, just pointing out that dynamic loading doesn't make a ton of sense in this scenario